### PR TITLE
Feature/dynamic routing

### PR DESCRIPTION
**Pull request by StianHaugland1**

### CODE REVIEW

File: h5p-bildetema-words-grid-view/src/components/TopicWordsGrid/TopicWordsGrid.tsx [hunk 1]
```diff
@@ -22,16 +22,19 @@ export const TopicWordsGrid: React.FC<TopicWordsGridProps> = ({
...
   React.useEffect(() => {
-    document
-      .getElementById(`toggle-${contentId}`)
-      ?.addEventListener("change", handleChange);
+    // Temp Hack: Navigating from topic -> words causes document.getElementById(`toggle-${contentId}`) to be null, if we not wait...
+    setTimeout(() => {
+      document
+        .getElementById(`toggle-${contentId}`)
+        ?.addEventListener("change", handleChange);
+    }, 10);
```
Comment (boyum) [inline h5p-bildetema-words-grid-view/src/components/TopicWordsGrid/TopicWordsGrid.tsx:30]: Will it work to wait one frame? All rendering is supposed to be done by then.

```suggestion
    requestAnimationFrame(() => {
      document
        .getElementById(`toggle-${contentId}`)
        ?.addEventListener("change", handleChange);
    });
```

Comment (boyum) [reply h5p-bildetema-words-grid-view/src/components/TopicWordsGrid/TopicWordsGrid.tsx:30]: Is there a reason why this is not handled by React?

File: h5p-bildetema/src/components/App/App.tsx [hunk 2]
```diff
@@ -1,18 +1,13 @@
...
 import { Bildetema } from "../Bildetema/Bildetema";
 
-type AppProps = {
-  currentLanguage: Language;
-};
-
 const queryClient = new QueryClient();
 
-export const App: React.FC<AppProps> = ({ currentLanguage }) => {
+export const App = (): JSX.Element => {
```
Comment (boyum) [inline h5p-bildetema/src/components/App/App.tsx:7]:
```suggestion
export const App: React.FC = () => {
```

File: h5p-bildetema/src/components/Bildetema/Bildetema.tsx [hunk 3]
```diff
@@ -47,243 +34,87 @@ export const defaultFavoriteLanguages: Language[] = [
...
 ];
 
-export const Bildetema: React.FC<BildetemaProps> = ({
-  // this variable can be used for setting default language
-  // when user first visits the page
-  // we can get the language object from languagesFromDB query data
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  currentLanguage: currentMetaLanguage,
-}) => {
+export const Bildetema = (): JSX.Element => {
```
Comment (boyum) [inline h5p-bildetema/src/components/Bildetema/Bildetema.tsx:37]:
```suggestion
export const Bildetema: React.FC = () => {
```

File: h5p-bildetema/src/components/Bildetema/Bildetema.tsx [hunk 4]
```diff
@@ -47,243 +34,87 @@ export const defaultFavoriteLanguages: Language[] = [
...
-  const navigate = useNavigate();
-  const location = useLocation();
+  const [topicIds, setTopicIds] = useState<TopicIds>({});
 
   const [topicsSize, setTopicsSize] = useState(TopicGridSizes.Big);
   const [isWordView, setIsWordView] = useState(false);
   const [showWrittenWords, setShowWrittenWords] = useState(true);
-  const [currentLanguage, setCurrentLanguage] = useState(
-    userData.currentLanguage ?? userData.favoriteLanguages[0],
+  const [favLanguages, setFavLanguages] = React.useState(
```
Comment (boyum) [inline h5p-bildetema/src/components/Bildetema/Bildetema.tsx:53]:
```suggestion
  const [favLanguages, setFavLanguages] = useState(
```

File: h5p-bildetema/src/components/Header/Header.tsx [hunk 5]
```diff
@@ -73,9 +74,28 @@ export const Header: React.FC<HeaderProps> = ({
...
 
-  const handleChangeLanguage = (newLanguage: Language): void => {
-    changeCurrentLanguage(newLanguage);
-    setUserData({ ...userData, currentLanguage: newLanguage });
+  const getLanguagePath = (language: Language): string => {
+    if (!topicId) return `/${language.code}`;
+
+    const topic = topicsFromDB?.find(el => el.id === topicId);
+    const topicWord = topic?.labelTranslations.get(language.code);
+    if (!topicWord) return `/${language.code}`;
```
Comment (boyum) [inline h5p-bildetema/src/components/Header/Header.tsx:82]: Should we point to the topic id if we can't find the word?

```suggestion
    if (!topicWord) return `/${language.code}/${topicId}`;
```

File: h5p-bildetema/src/components/Header/Header.tsx [hunk 6]
```diff
@@ -73,9 +74,28 @@ export const Header: React.FC<HeaderProps> = ({
...
+    const topicPath =
+      topicWord.label !== ""
+        ? labelToUrlComponent(topicWord.label)
+        : labelToUrlComponent(topicWord.id);
+    if (!subTopicId) return `/${language.code}/${topicPath}`;
+
+    const subTopicWord = topic?.subTopics
+      .get(subTopicId)
+      ?.labelTranslations.get(language.code);
+    if (!subTopicWord) return `/${language.code}/${topicPath}`;
```
Comment (boyum) [inline h5p-bildetema/src/components/Header/Header.tsx:93]: As abpve

```suggestion
    if (!subTopicWord) return `/${language.code}/${topicPath}/${subTopicId}`;
```

File: h5p-bildetema/src/components/Header/Header.tsx [hunk 7]
```diff
@@ -104,16 +121,13 @@ export const Header: React.FC<HeaderProps> = ({
...
-            setUserData={setUserData}
             favLanguages={favLanguages}
             setFavLanguages={setFavLanguages}
-            handleChangeLanguage={handleChangeLanguage}
           />
         </div>
       </div>
       <div className={styles.bottom}>
-        <Breadcrumbs currentLanguageCode={currentLanguage.code} />
+        <Breadcrumbs currentLanguageCode="nob" />
```
Comment (boyum) [inline h5p-bildetema/src/components/Header/Header.tsx:130]: Should we revert this change?

File: h5p-bildetema-words-grid-view/src/components/TopicWordsGrid/TopicWordsGrid.tsx (hunk 1 above)
Comment (StianHaugland1) [reply h5p-bildetema-words-grid-view/src/components/TopicWordsGrid/TopicWordsGrid.tsx:30]: This was added because when it was handled by react and we toggled the showWrittenWords button it caused all the images to "blink". (guess they reloaded)

**Review by boyum — APPROVED** [review]
